### PR TITLE
JS API docs for signing service

### PR DESCRIPTION
**Pull request by coder11**

### CODE REVIEW

File: fluence-js/3_in_depth.md [hunk 1]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
 To learn more about tetraplets and application security see [Security](https://github.com/fluencelabs/gitbook-docs/tree/77344eb147c2ce17fe1c0f37013082fc85c1ffa3/js-sdk/knowledge_security.md)
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:350]: link is dead

File: fluence-js/3_in_depth.md [hunk 2]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
 
 To see full specification of `CallParams` type see [API reference](https://github.com/fluencelabs/gitbook-docs/tree/77344eb147c2ce17fe1c0f37013082fc85c1ffa3/js-sdk/js-sdk/6_reference/modules.md)
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:352]: link is dead

File: fluence-js/3_in_depth.md [hunk 3]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+
+### Signing service
+
+Signing service is useful when you want to sign arbitrary data and pass it further inside a singe aqua script. Signing service allows to restrict it's usage for security reasons: e.g you don't want to sign anything except it comes from a trusted source. The aqua side API is the following:
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:356]: singe -> single
What's the significance of 'single' aqua script?
restrict it's usage  -> its; it's as in the arbitrary data ?
e.g you don't want to sign anything except <if/when > it comes from a trusted source. -> don't quite follow: data comes from a trusted source, then i sign it, then it's used in an aqua script ... ??

File: fluence-js/3_in_depth.md [hunk 4]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+data SignResult:
+    -- Was call successful or not
+    success: bool
+    -- Error message. Will be null if the call is successful
+    error: ?string
+    -- Signature as byte array. Will be null if the call is not successful
+    signature: ?[]u8
+
+service Sig("sig"):
+    -- Signs data with the service's private key.
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:368]: which service?

File: fluence-js/3_in_depth.md [hunk 5]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+    -- Depending on implementation the service might check call params to restrict usage for security reasons.
+    -- By default it is only allowed to be used on the same peer the particle was initiated
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:370]: it -> referring to the/a service ?

File: fluence-js/3_in_depth.md [hunk 6]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+    --   registry.get_key_bytes
+    --   registry.get_record_bytes
+    -- Argument: data - byte array to sign
+    -- Returns: signature as SignResult structure
+    sign(data: []u8) -> SignResult
+
+    -- Given the data and signature both as byte arrays, returns true if the signature is correct, false otherwise.
+    verify(signature: []u8, data: []u8) -> bool
+
+    -- Gets the service's public key of.
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:383]: of.  -> ??

File: fluence-js/3_in_depth.md [hunk 7]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+    get_pub_key() -> string
+
+```
+
+FluenceJS ships the service implementation as the JavaScript class:
+
+```typescript
+/**
+ * A predicate of call params for sig service's sign method which determines whether signing operation is allowed or not
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:392]: there's got to be a better way to express the notion of signing auth

File: fluence-js/3_in_depth.md [hunk 8]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+
+export class Sig implements SigDef {
+  private _keyPair: KeyPair;
+
+  constructor(keyPair: KeyPair) {
+    this._keyPair = keyPair;
+  }
+
+  /**
+   * A predicate of call params for sig service's sign method which determines whether signing operation is allowed or not
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:404]: dupe from above

File: fluence-js/3_in_depth.md [hunk 9]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+    callParams: CallParams<"data">
+  ): Promise<SignResult> {
+    // ...
+  }
+
+  /**
+   * Verifies the signature. Required by aqua
+   */
+  verify(signature: number[], data: number[]): Promise<boolean> {
+    // ...
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:429]: is there some content coming or is that it?

File: fluence-js/3_in_depth.md [hunk 10]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+const customSig = new Sig(customKeyPair);
+
+// restrict sign usage to our needs
+customSig.securityGuard = allowServiceFn("my_service", "my_function");
+
+// register the service. Please note, that service id ("CustomSig" here) has to be specified.
+registerSig("CustomSig", customSig);
+```
+
+for a non-default peer, the instance has to be specified:
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:489]: what's a non-default peer? link to definition would help

File: fluence-js/3_in_depth.md [hunk 11]
```diff
@@ -350,3 +350,177 @@ Tetraplets have the form of:
...
+```typescript
+const peer = new FluencePeer();
+await peer.start();
+
+// ...
+
+registerSig(peer, "CustomSig", customSig);
+```
+
+FluencePeer is shipped with the default signing service implementation, registered with id "Sig". Is is useful to work with Trust Graph and Registry API. The default service provides the following restrictions on the `sign` operation:
```
Comment (boneyard93501) [inline fluence-js/3_in_depth.md:500]: FluencePeer is ships with the default signing service registered with id "Sig".
TrustGraph 
default service is ?

**Comment by coder11** [comment]
@boneyard93501  really need these docs for the hackathon. Will address all the comments later

**Pull request merged**